### DISCUSSION TITLE
Decrease flash usage by disabling some peripherals?

### DISCUSSION THREAD

**Discussion by China-Mr-Hou** [stm32duino/Arduino_Core_STM32#2296] [Q&A] [answered]
Hi, everyone. I encountered flash memory overflow with some "large" codes on STM32F103C6. This project uses UARTs only (no I2C, SPI, etc.). Would it be able to disable some unused peripherals? I saw every peripheral HAL library was compiled and linked into the final elf/bin file. Even an empty program (only setup and loop function) takes up quite a bit of the flash memory.

**Comment by fpistm** [comment, marked as answer]
Hi @China-Mr-Hou 
You can refer to:
https://github.com/stm32duino/Arduino_Core_STM32/wiki/HAL-configuration#example-of-blinkino-for-nucleo-l031k6-with-32768-bytes-of-flash-and-8192-bytes-of-ram
And also:
https://github.com/stm32duino/Arduino_Core_STM32/wiki/Custom-definitions#custom-pinmap-array

**Reply by China-Mr-Hou** [reply]
Hello @fpistm , I need more help. I was trying the method given by creating a header named hal_conf_extra.h and added these code below.
```c
#ifndef _HAL_CONF_EXTRA_H_
#define _HAL_CONF_EXTRA_H_

#define HAL_I2C_MODULE_DISABLED
#define HAL_RTC_MODULE_DISABLED
#define HAL_SPI_MODULE_DISABLED
#define HAL_ETH_MODULE_DISABLED
#define HAL_SD_MODULE_DISABLED
#define HAL_QSPI_MODULE_DISABLED

#endif
``` 
The file was added to my FOC code that I slightly mentioned in #2287 since it's very small and can be successfully compiled anytime. However, whether  hal_conf_extra.h exist or not doesn't change the use of flash (in my case 40800 bytes on G474RE). I also tried hal_conf_custom.h method and got error message flooding from the compiler which I believe that's not the normal way to handle this problem. btw, the wiki you provided says DAC module is not enabled by default, but I still saw stm32yyxx_hal_dac.c and stm32yyxx_ll_dac.c being compiled and the corresponding .o file generated. 

Update after posting 5mins: Does stm32duino core auto detects whether a specialised peripheral is used or not? I added pinMode(pin, INPUT); and analogRead(pin); by design (HAL_ADC_MODULE_DISABLED is not added from the very beginning to just now), and the usage of flash increased to 47304 bytes, which is approx. 6kB more.

**Reply by fpistm** [reply]
I guess you have a cache issue. Ensure to have all rebuild (by deleting the Arduino tmp file or by changing an options in the menu, ex upload method).
You will always see all files build and .o generated, as files are always present, simply the code are guard by the HAL define so if not defined, the code is not build.
The code does not detect if a peripheral is used or not but at link stage is a function is not used it is simply stripped.

**Reply by China-Mr-Hou** [reply]
I don't think it was cache that matter. Originally I was coding in VSCode with PlatformIO and it also has stm32duino core under Arduino Framework. I tried cleaning cached files and it didn't seem to work. I also tried Arduino IDE version 2.3.2 just now and followed your instructions with a blink sketch and a blank sketch, however the same result appeared. I personally think that I misused hal_conf_extra.h

**Reply by fpistm** [reply]
Here my result for a Nucleo F103RB:

<details>
  <summary>Blink + Serial enabled and used in the sketch:</summary>

```C++
// the setup function runs once when you press reset or power the board
void setup() {
  // initialize digital pin LED_BUILTIN as an output.
  pinMode(LED_BUILTIN, OUTPUT);
#ifdef HAL_UART_MODULE_ENABLED
  Serial.begin(9600);
#endif
}

// the loop function runs over and over again forever
void loop() {
  digitalWrite(LED_BUILTIN, HIGH);  // turn the LED on (HIGH is the voltage level)
  delay(1000);                      // wait for a second
  digitalWrite(LED_BUILTIN, LOW);   // turn the LED off by making the voltage LOW
  delay(1000);                      // wait for a second
#ifdef HAL_UART_MODULE_ENABLED
  Serial.println("loop");
#endif
}
```
</details>

> Sketch uses 12540 bytes (9%) of program storage space. Maximum is 131072 bytes.
> Global variables use 1132 bytes (5%) of dynamic memory, leaving 19348 bytes for local variables. Maximum is 20480 bytes.

<details>
  <summary>Blink + Serial enabled + hal_conf_extra.h:</summary>

` hal_conf_extra.h` contains:
```C++
#ifndef _HAL_CONF_EXTRA_H_
#define _HAL_CONF_EXTRA_H_

#define HAL_ADC_MODULE_DISABLED
#define HAL_I2C_MODULE_DISABLED
#define HAL_RTC_MODULE_DISABLED
#define HAL_SPI_MODULE_DISABLED
#define HAL_TIM_MODULE_DISABLED

#endif
```
</details>

> Sketch uses 9424 bytes (7%) of program storage space. Maximum is 131072 bytes.
> Global variables use 684 bytes (3%) of dynamic memory, leaving 19796 bytes for local variables. Maximum is 20480 bytes.
> 

<details>
  <summary>Blink + Serial enabled + hal_conf_extra.h + PinMap_UART_* redefined:</summary>

Added on top of the sketch:
```C++
#ifdef HAL_UART_MODULE_ENABLED
const PinMap PinMap_UART_TX[] = {
  {PA_2,  USART2, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART2_DISABLE)},
  {NC,    NP,     0}
};

const PinMap PinMap_UART_RX[] = {
  {PA_3,  USART2, STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_USART2_DISABLE)},
  {NC,    NP,     0}
};

const PinMap PinMap_UART_RTS[] = {
  {NC,         NP,     0}
};

const PinMap PinMap_UART_CTS[] = {
  {NC,         NP,     0}
};
#endif
```
</details>

> Sketch uses 9236 bytes (7%) of program storage space. Maximum is 131072 bytes.
> Global variables use 684 bytes (3%) of dynamic memory, leaving 19796 bytes for local variables. Maximum is 20480 bytes.


Blink + hal_conf_extra.h + + Serial disabled:

> Sketch uses 4756 bytes (3%) of program storage space. Maximum is 131072 bytes.
> Global variables use 392 bytes (1%) of dynamic memory, leaving 20088 bytes for local variables. Maximum is 20480 bytes.
> 


So there is no issue with `hal_conf_extra.h`, as stated I really think you have a cache issue but can't help on this.
